hoot-services/src/main/java/hoot/services/job/JobStatusManagerImpl.java

```diff
@@ -97,19 +105,71 @@ public void addJob(Job job) {
+                .from(jobStatus)
+                .where(jobStatus.jobId.eq(jobId).and(jobStatus.userId.eq(userId)))
+                .fetchFirst();
+
             //hack to check that user owns this job before deleting commands
             //no cascade delete in querydsl
             if (js != null) {
 
                 createQuery().delete(commandStatus)
-                    .where(commandStatus.jobId.eq(jobId)).execute();
+                    .where(commandStatus.jobId.eq(jobId))
+                    .execute();
 
                 createQuery().delete(jobStatus)
-                    .where(jobStatus.jobId.eq(jobId).and(jobStatus.userId.eq(userId))).execute();
+                    .where(jobStatus.jobId.eq(jobId).and(jobStatus.userId.eq(userId)))
+                    .execute();
+
+                File workDir;
+                Integer jobType = js.getJobType();
+
+                // Clean up left over data related to specific jobs
+                if (jobType == JobType.DERIVE_CHANGESET.ordinal() || jobType == JobType.UPLOAD_CHANGESET.ordinal()) {
+                    List<String> fileFilters;
+
+                    if (jobType == JobType.DERIVE_CHANGESET.ordinal()) {
+                        // these are the files that can be produced from a derive changeset job
+                        fileFilters = Arrays.asList("diff.osc", "diff.tags.osc", "reference.osm", "secondary.osm", "stats.json", "stats.tags.json");
+
+                        workDir = new File(CHANGESETS_FOLDER, js.getJobId());
+                    } else {
+                        // these are the files that can be produced from a upload changeset job
+                        fileFilters = Arrays.asList("diff-error.osc", "diff-remaining.osc");
+
+                        // An upload jobs work directory is actually the parents(derive changeset) directory
+                        Map<String, String> tags = PostgresUtils.postgresObjToHStore(js.getTags());
+                        workDir = new File(CHANGESETS_FOLDER, tags.get("parentId"));
+                    }
+
+
+                    if (workDir.exists()) {
+                        // Only delete the files that are in the fileFilters list
+                        File[] files = workDir.listFiles(pathname -> fileFilters.contains(pathname.getName()));
+
+                        if (files != null && files.length > 0) {
+                            for(File file : files) {
+                                file.delete();
+                            }
+                        }
+
+                        // only delete if empty.
+                        // this is due to a case where if user deletes a derive changeset, we still want to keep upload changeset files if they exist. if not delete. Same for deleting an upload job if the changeset job still exists
+                        if (workDir.list().length == 0) {
+                            FileUtils.deleteDirectory(workDir);
+                        }
+                    }
```

Comment:
the special case i found:

user runs derive changeset and upload

if user deletes the derive job, we still want them to be able to 'download conflicted changes' for the upload job.

if user deletes upload job, still want them to be able to 'download changeset' for the derive changeset job

Reply:
Yep, this handles that well